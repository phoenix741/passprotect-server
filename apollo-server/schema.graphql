scalar Date

# Encrypted content
type EncryptedContent {
  # Content encrypted with salt
  content: String!
  # Authentification tag
  authTag: String!
}

# Encrypted content
input EncryptedContentInput {
  # Content encrypted with salt
  content: String!
  # Authentification tag
  authTag: String!
}

type Error {
  fieldName: String,
  message: String
}

type Errors {
  errors: [Error]!
}

# Master key encrypted for the user
type EncryptedMasterKey {
  # Salt added to encrypt the master key
  salt: String!
  # Encrypted master key
  encryptedKey: EncryptedContent!
}

# Master key encrypted for the user
input EncryptedMasterKeyInput {
  # Salt added to encrypt the master key
  salt: String!
  # Encrypted master key
  encryptedKey: EncryptedContentInput!
}

# User defined in the application
type User {
  # Login of the user
  _id: ID!,
  # Encrypted master key
  encryption: EncryptedMasterKey!,
  # Creation date of the user
  creationAt: Date,
  # Role of the user (can be user or admin)
  role: String
  # Wallet lines of the user.
  lines: [WalletLine]!
}

input RegistrationUser {
  # Login of the user
  _id: ID!,
  # Password used for registration
  password: String!
  # Encrypted master key
  encryption: EncryptedMasterKeyInput!,
}

union RegisterUserResult = User | Errors

# Wallet line content encrypted
type EncryptedWalletLine {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContent!
}

# Wallet line content encrypted
input EncryptedWalletLineInput {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContentInput!
}

# Wallet
type WalletLine {
  # Id of the line
  _id: ID!
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLine!,
  # Owner of the list
  user: User!,
  # Last updated date of the line
  updatedAt: Date,
  # Revision (modification count) of the line
  _rev: Int
}

# Wallet line used as input
input WalletLineInput {
  # Id of the line, if not filled application create a new line
  _id: ID
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLineInput!,
  # Revision (modification count) of the line, used to ensure modification of the last line
  _rev: Int
}

union CreateUpdateLineResult = WalletLine | Errors

# Information needed to connect to the server and retreive object from database.
input ConnectionInformationInput {
  # Login of the user
  username: String!,
  # Password used for connection
  password: String!
}

# Result of the connection
type ConnectionInformation {
  # Token to be used in connection information
  token: String,
  # The connected user
  user: User!
}

union CreateSessionResult = ConnectionInformation | Errors

# Wallet transaction, used to synchronize lines between devices
type WalletTransaction {
  # Id of the transaction
  _id: ID!,
  # Type of the object modified (only line)
  type: String!,
  # Line
  line: WalletLine,
  # Owner of the transaction
  user: User!,
  # The line before the modification
  before: WalletLine,
  # The line after the modification
  after: WalletLine,
  # The modification date
  updatedAt: Date!,
  # A SHA-512 calculate on the modification.
  sha512: String!
}

type RootQuery {
  # Request a user
  user(id: ID!): User,

    # Request the list of lines from the connected users.
  lines: [WalletLine],

  # Request a wallet line
  line(id: ID!): WalletLine,

	# Request all groups available on lines
	groups: [String!]

  # Request the current session
  session: User

  # Request all transactions that happen since the last synchronisation date.
  transactions(earliest: Date!): [WalletTransaction]
}

type RootMutation {
  registerUser(input: RegistrationUser!): RegisterUserResult

  createUpdateLine(input: WalletLineInput!): CreateUpdateLineResult
  removeLine(id: ID): Errors

  createSession(input: ConnectionInformationInput!): CreateSessionResult
}

type RootSubscription {
  transactionAdded: WalletTransaction!
}

schema {
  query: RootQuery
  mutation: RootMutation
  subscription: RootSubscription
}
