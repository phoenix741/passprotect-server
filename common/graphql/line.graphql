# Wallet line content encrypted
type EncryptedWalletLine {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContent!
}

# Wallet line content encrypted
input EncryptedWalletLineInput {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContentInput!
}

# Wallet
type WalletLine {
  # Id of the line
  _id: ID!
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLine!,
  # Owner of the list
  user: User!,
  # Last updated date of the line
  updatedAt: Date,
  # Revision (modification count) of the line
  _rev: Int
}

# Wallet line used as input
input WalletLineInput {
  # Id of the line, if not filled application create a new line
  _id: ID
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLineInput!,
  # Revision (modification count) of the line, used to ensure modification of the last line
  _rev: Int
}

extend type RootQuery {
  # Request the list of lines from the connected users.
  lines: [WalletLine],

  # Request a wallet line
  line(id: ID!): WalletLine,

	# Request all groups available on lines
	groups: [String!]
}

union CreateUpdateLineResult = WalletLine | Errors

extend type RootMutation {
  createUpdateLine(input: WalletLineInput!): CreateUpdateLineResult
  removeLine(id: ID): Errors
}
