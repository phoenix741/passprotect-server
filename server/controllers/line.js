'use strict';

const _ = require('underscore');
const router = require('express-promise-router')();
const i18n = require('i18next');
const authenticate = require('../utils/passport').authenticate;
const permission = require('../utils/passport').permission;

module.exports = function (app, services) {
	/**
	 * @apiDefine text Informations to encrypt (client side) for text
	 */
	/**
	 * @apiDefine password Informations to encrypt (client side) for password
	 */
	/**
	 * @apiDefine card Informations to encrypt (client side) for payment card
	 */

	/**
	 * @apiDefine UpdateLine
	 *
	 * @apiParam {String="text","card","password"} type type of line
	 * @apiParam {String} label label of the line
	 * @apiParam {Object} encryption Informations used for encrypted data
	 * @apiParam {Object} encryption.informations Encrypted informations
	 * @apiParam {String} encryption.informations.authTag Authentification tag of the encrypted information
	 * @apiParam {String} encryption.informations.content Encrypted informations
	 * @apiParam {String} encryption.salt Salt used to encrypt/decrypt informations
	 *
	 * @apiParam (text) {String} text Text that will be encrypted
	 * @apiParam (text) {String} notes Custom note added to the line
	 *
	 * @apiParam (password) {String} username Username associated to the pasword
	 * @apiParam (password) {String} password The password
	 * @apiParam (password) {String} siteUrl The site where the password is used
	 * @apiParam (password) {String} notes Custom note added to the line
	 *
	 * @apiParam (card) {String="VISA","Carte bleu","Mastercard","American Express","Maestro"} type Type of card
	 * @apiParam (card) {String} nameOnCard Name of the person on the card
	 * @apiParam (card) {String} cardNumber The card number
	 * @apiParam (card) {String} cvv The code in the back on the card
	 * @apiParam (card) {String} expiry The expiration date
	 * @apiParam (card) {String} code The code to use to pay with the card
	 * @apiParam (card) {String} notes Custom note added to the line
	 *
	 * @apiParamExample {json} Line:
	 *   {
	 *     "type": "text",
	 *     "label": "My custom text",
	 *     "encryption": {
	 *       "salt": "83b01169a90fe764",
	 *       "informations": {
	 *         "content": "d13a32829c2547b8bca6da5bb568d3",
	 *         "authTag": "db47515bdcb7643484d60189b7ad4e33"
	 *       }
	 *     }
	 *   }
	 */

	/**
	 * @apiDefine ResponseLineComplete
	 *
	 * @apiSuccess {String} _id Id of the line
	 * @apiSuccess {String} type type of line
	 * @apiSuccess {String} label label of the line
	 * @apiSuccess {Object} encryption Informations used for encrypted data
	 * @apiSuccess {Object} encryption.informations Encrypted informations
	 * @apiSuccess {String} encryption.informations.content Encrypted content of the line
	 * @apiSuccess {String} encryption.informations.authTag Authentification tag of the encrypted informations
	 * @apiSuccess {String} encryption.salt Salt used to encrypt informations
	 *
	 *
	 * @apiSuccessExample {json} Success text:
	 *   HTTP/1.1 200 OK
	 *     {
	 *       "_id": "57685936f4495bc62c4d28a3",
	 *       "type": "text",
	 *       "label": "My custom text",
	 *       "encryption": {
	 *         "salt": "83b01169a90fe764",
	 *         "informations": {
	 *           "content": "d13a32829c2547b8bca6da5bb568d3",
	 *           "authTag": "db47515bdcb7643484d60189b7ad4e33"
	 *         }
	 *       }
	 *     }
	 */

	router.route('')
	/**
	 * @api {get} /api/lines Request all lines
	 *
	 * @apiDescription Request the list of lines from the connected users. Only the _id, type, and label of the line is
	 * given. No encrypted information are send in this method.
	 *
	 * @apiName GetLines
	 * @apiGroup Line
	 * @apiPermission ROLE_USER
	 *
	 * @apiSuccess {String} _id Id of the line
	 * @apiSuccess {String} type type of line
	 * @apiSuccess {String} label label of the line
	 *
	 * @apiSuccessExample Success:
	 *     HTTP/1.1 200 OK
	 *     [
	 *       {
	 *         "_id": "57685936f4495bc62c4d28a3",
	 *         "type": "card",
	 *         "label": "My credit card"
	 *       },
	 *       {
	 *         "_id": "57685936f4495bc62c4d28a4",
	 *         "type": "password",
	 *         "label": "My password"
	 *       },
	 *     ]
	 *
	 * @apiError Unauthorized Only the connected user can access to list of lines
	 * @apiError Forbidden Your are not allowed to access to this API
	 *
	 * @apiErrorExample {json} Unauthorized:
	 *     HTTP/1.1 401 Unauthorized
	 *
	 * @apiErrorExample {json} Forbidden:
	 *     HTTP/1.1 403 Forbidden
	 */
		.get((req, res) => {
			const offset = parseInt(req.query.offset || 0);
			const limit = parseInt(req.query.limit || 10);

			return services.line.getLines(req.user, {offset, limit}).map(filterLine).then(res.json.bind(res));
		})

		/**
		 * @api {post} /api/lines Create new line
		 * @apiName PostLines
		 * @apiGroup Line
		 * @apiPermission ROLE_USER
		 *
		 * @apiUse UpdateLine
		 * @apiUse ResponseLineComplete
		 *
		 * @apiError Unauthorized Only the connected user can access to list of lines
		 *
		 * @apiErrorExample {json} Unauthorized:
		 *     HTTP/1.1 401 Unauthorized
		 */
		.post((req, res) => {
			const data = _.pick(req.body, 'type', 'label', 'encryption');

			return saveLine(data, req, res);
		});

	router.param('lineId', (req, res, next, lineId) => {
		services.line.getLine(lineId, req.user).then(line => {
			// Can't access or update lines that isn't associated to the account of the user
			if (line.user !== req.user._id) {
				const notAuthorizedError = new Error('Not authorized to access to thhis subscriptions');
				notAuthorizedError.status = 403;
				return Promise.reject(notAuthorizedError);
			}

			req.line = line;
			return next();
		}).catch(next);
	});

	/**
	 * @api {get} /api/lines/:lineId Request line
	 * @apiName GetLine
	 * @apiGroup Line
	 * @apiPermission ROLE_USER
	 *
	 * @apiParam {String} lineId The id of the line
	 *
	 * @apiUse ResponseLineComplete
	 *
	 * @apiError Unauthorized Only the connected user can access to list of lines
	 * @apiError Forbidden Your are not allowed to access to this API
	 *
	 * @apiErrorExample {json} Unauthorized:
	 *     HTTP/1.1 401 Unauthorized
	 *
	 * @apiErrorExample {json} Forbidden:
	 *     HTTP/1.1 403 Forbidden
	 */
	router.route('/:lineId')
		.get((req, res) => res.json(filterLine(req.line)))
		/**
		 * @api {put} /api/lines/:lineId Update the line
		 * @apiName PutLine
		 * @apiGroup Line
		 * @apiPermission ROLE_USER
		 *
		 * @apiUse UpdateLine
		 * @apiUse ResponseLineComplete
		 */
		.put((req, res) => {
			const data = _.pick(req.body, 'type', 'label', 'encryption');

			// Force the id of the line to ensure that the user doesn't try to change it
			data._id = req.line._id;

			return saveLine(data, req, res);
		})
		/**
	 	 * @api {delete} /api/lines/:lineId Delete the line
		 * @apiName DeleteLine
		 * @apiGroup Line
		 * @apiPermission ROLE_USER
		 *
		 * @apiSuccessExample {json} Success-Response:
		 *     HTTP/1.1 204 OK
	 	 */
		.delete((req, res) => {
			return services.line.removeLine(req.line._id).then(function() {
				res.status(204).send();
			});
		});

	app.use('/api/lines', authenticate(), permission(), router);

	/**
	 * Filter information that the user is authorized to view in the response
	 * @param {Object} line The response
	 * @returns {Object} The filtered object.
	 */
	function filterLine(line) {
		return _.pick(line, '_id', 'type', 'label', 'encryption');
	}

	/**
	 * Create or modify a line given by the request.
	 * @param {Object} data Object to create or modify
	 * @param {Object} req Request use to read the user
	 * @param {Object} res Response used to send the response.
	 * @returns {Promise} Promise that the line will be saved.
	 */
	function saveLine(data, req, res) {
		// Force the user stored in the line.
		data.user = req.user._id;

		const validationError = new Error();
		validationError.status = 400;

		const typeChoices = ['card', 'password', 'text'];
		if (!_.contains(typeChoices, data.type)) {
			validationError.message = i18n.t('error:line.400.type', {choices: typeChoices.join(', ')});
			return Promise.reject(validationError);
		}

		if (!_.isString(data.label) || _.isEmpty(data.label)) {
			validationError.message = i18n.t('error:line.400.label');
			return Promise.reject(validationError);
		}

		return services.line.saveLine(data, req.user).then(function (line) {
			res.json(filterLine(line));
		});
	}
};
