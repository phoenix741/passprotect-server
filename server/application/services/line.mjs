'use strict';

import debug from 'debug';
import {NotFoundError} from '../models/exception';
import {getLines as getLinesModel, getLine as getLineModel, saveLine as saveLineModel, removeLine as removeLineModel} from '../models/line';
import {createTransaction} from '../services/transaction';

const log = debug('App:Service:Line');

export function getLines(user) {
	log('Get all lines from user ', user._id);

	const filter = {};
	filter.user = user._id;

	const sort = {};
	sort.group = 1;
	sort.label = 1;

	return getLinesModel(filter, sort);
}

export function getLine(id) {
	log(`Get the line with id ${id}`);
	return getLineModel(id);
}

export function saveLine(line) {
	log(`Create the line of type ${line.type}`);

	line.updatedAt = new Date();

	return getLineIfAvailable(line._id, line._rev)
		.then(function (oldLine) {
			return saveLineModel(line).tap(newLine => createTransaction('line', oldLine, newLine));
		});
}

export function removeLine(id) {
	log(`Remove the line with the id ${id}`);

	return getLineIfAvailable(id)
		.tap(() => removeLineModel(id))
		.tap(oldLine => createTransaction('line', oldLine));
}

function getLineIfAvailable(id, rev) {
	return getLineModel(id, rev).catch(NotFoundError, () => null);
}
