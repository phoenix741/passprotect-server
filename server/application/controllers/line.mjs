import {omit, pick, includes, isString, isEmpty} from 'lodash'
import expressPromiseRouter from 'express-promise-router'
import i18n from 'i18next'
import fs from 'fs'
import path from 'path'
import debug from 'debug'
import {authenticate, permission, checkPermission} from '../utils/passport'
import {getLines, getLine, removeLine, saveLine} from '../services/line'
import {getGroups} from '../services/group'

const log = debug('App:Controllers:Line')

const router = expressPromiseRouter()
export default [authenticate(), permission(), router]

log('Load line type definition')
export const typeDefs = [
  fs.readFileSync(path.join(__dirname, '..', '..', '..', 'common', 'graphql', 'line.graphql'), 'utf-8')
]

export const resolvers = {
  RootQuery: {
    lines (obj, args, {user}) {
      return checkPermission(user).then(() => getLines(user)).map(line => checkPermission(user, [], line.user).return(line)).map(filterLine)
    },

    line (obj, {id}, {user}) {
      return getLine(id).tap(line => checkPermission(user, [], line.user)).then(filterLine)
    },

    groups (obj, args, {user}) {
      return checkPermission(user).then(() => getGroups(user))
    }
  },
  User: {
    lines (obj, args, {user}) {
      return getLines(obj._id).map(line => checkPermission(user, [], line.user).return(line)).map(filterLine)
    }
  },

  RootMutation: {
    createUpdateLine (obj, {input}, {user}) {
      let checkPermissionPromise = Promise.resolve(input)
      if (input._id) {
        checkPermissionPromise = getLine(input._id)
          .tap(line => checkPermission(user, [], line.user))
          .then(line => {
            input._id = line._id
            return input
          })
      }

      input.user = user._id
      return checkPermissionPromise
        .then(input => sanitizeInput(input))
        .then(data => saveLine(data))
        .then(filterLine)
        .catch(parseErrors)
    },

    removeLine (obj, {id}, {user}) {
      return getLine(id)
        .tap(line => checkPermission(user, [], line.user))
        .then(() => removeLine(id))
        .then(() => ({errors: []}))
        .catch(parseErrors)
    }
  },

  CreateUpdateLineResult: {
    __resolveType (obj) {
      if (obj.errors) {
        return 'Errors'
      }
      return 'WalletLine'
    }
  }
}

/**
 * @apiDefine text Informations to encrypt (client side) for text
 */
/**
 * @apiDefine password Informations to encrypt (client side) for password
 */
/**
 * @apiDefine card Informations to encrypt (client side) for payment card
 */

/**
 * @apiDefine UpdateLine
 *
 * @apiParam {String="text","card","password"} type type of line
 * @apiParam {String} label label of the line
 * @apiParam {Object} encryption Informations used for encrypted data
 * @apiParam {Object} encryption.informations Encrypted informations
 * @apiParam {String} encryption.informations.authTag Authentification tag of the encrypted information
 * @apiParam {String} encryption.informations.content Encrypted informations
 * @apiParam {String} encryption.salt Salt used to encrypt/decrypt informations
 *
 * @apiParam (text) {String} text Text that will be encrypted
 * @apiParam (text) {String} notes Custom note added to the line
 *
 * @apiParam (password) {String} username Username associated to the pasword
 * @apiParam (password) {String} password The password
 * @apiParam (password) {String} siteUrl The site where the password is used
 * @apiParam (password) {String} notes Custom note added to the line
 *
 * @apiParam (card) {String="VISA","Carte bleu","Mastercard","American Express","Maestro"} type Type of card
 * @apiParam (card) {String} nameOnCard Name of the person on the card
 * @apiParam (card) {String} cardNumber The card number
 * @apiParam (card) {String} cvv The code in the back on the card
 * @apiParam (card) {String} expiry The expiration date
 * @apiParam (card) {String} code The code to use to pay with the card
 * @apiParam (card) {String} notes Custom note added to the line
 *
 * @apiParamExample {json} Line:
 *   {
 *     "type": "text",
 *     "label": "My custom text",
 *     "encryption": {
 * )      "salt": "83b01169a90fe764",
 *       "informations": {
 *         "content": "d13a32829c2547b8bca6da5bb568d3",
 *         "authTag": "db47515bdcb7643484d60189b7ad4e33"
 *       }
 *     }
 *   }
 */

/**
 * @apiDefine ResponseLineComplete
 *
 * @apiSuccess {String} _id Id of the line
 * @apiSuccess {String} type type of line
 * @apiSuccess {String} label label of the line
 * @apiSuccess {Object} encryption Informations used for encrypted data
 * @apiSuccess {Object} encryption.informations Encrypted informations
 * @apiSuccess {String} encryption.informations.content Encrypted content of the line
 * @apiSuccess {String} encryption.informations.authTag Authentification tag of the encrypted informations
 * @apiSuccess {String} encryption.salt Salt used to encrypt informations
 *
 *
 * @apiSuccessExample {json} Success text:
 *   HTTP/1.1 200 OK
 *     {
 *       "_id": "57685936f4495bc62c4d28a3",
 *       "type": "text",
 *       "label": "My custom text",
 *       "encryption": {
 *         "salt": "83b01169a90fe764",
 *         "informations": {
 *           "content": "d13a32829c2547b8bca6da5bb568d3",
 *           "authTag": "db47515bdcb7643484d60189b7ad4e33"
 *         }
 *       }
 *     }
 */

router.route('')
  /**
   * @api {get} /api/lines Request all lines
   *
   * @apiDescription Request the list of lines from the connected users. Only the _id, type, and label of the line is
   * given. No encrypted information are send in this method.
   *
   * @apiName GetLines
   * @apiGroup Line
   * @apiPermission ROLE_USER
   *
   * @apiParam {Number} [offset=0] Offset for pagination
   * @apiParam {Number} [limit=10000] Limit of items used for pagination
   *
   * @apiSuccess {String} _id Id of the line
   * @apiSuccess {String} type type of line
   * @apiSuccess {String} label label of the line
   *
   * @apiSuccessExample Success:
   *     HTTP/1.1 200 OK
   *     [
   *       {
   *         "_id": "57685936f4495bc62c4d28a3",
   *         "type": "card",
   *         "label": "My credit card"
   *       },
   *       {
   *         "_id": "57685936f4495bc62c4d28a4",
   *         "type": "password",
   *         "label": "My password"
   *       },
   *     ]
   *
   * @apiError Unauthorized Only the connected user can access to list of lines
   * @apiError Forbidden Your are not allowed to access to this API
   *
   * @apiErrorExample {json} Unauthorized:
   *     HTTP/1.1 401 Unauthorized
   *
   * @apiErrorExample {json} Forbidden:
   *     HTTP/1.1 403 Forbidden
   */
  .get((req, res) => {
    return getLines(req.user).map(filterLine).then(res.json.bind(res))
  })

  /**
   * @api {post} /api/lines Create new line
   * @apiName PostLines
   * @apiGroup Line
   * @apiPermission ROLE_USER
   *
   * @apiUse UpdateLine
   * @apiUse ResponseLineComplete
   *
   * @apiError Unauthorized Only the connected user can access to list of lines
   *
   * @apiErrorExample {json} Unauthorized:
   *     HTTP/1.1 401 Unauthorized
   */
  .post((req, res) => {
    const data = omit(req.body, '_id', 'user')

    return checkAndSaveLine(data, req, res)
  })

router.param('lineId', (req, res, next, lineId) => {
  getLine(lineId, req.user)
    .then(line => {
      // Can't access or update lines that isn't associated to the account of the user
      if (line.user !== req.user._id) {
        const notAuthorizedError = new Error('Not authorized to access to thhis subscriptions')
        notAuthorizedError.status = 403
        return Promise.reject(notAuthorizedError)
      }

      req.line = line
      return next()
    })
    .catch(next)
})

/**
 * @api {get} /api/lines/:lineId Request line
 * @apiName GetLine
 * @apiGroup Line
 * @apiPermission ROLE_USER
 *
 * @apiParam {String} lineId The id of the line
 *
 * @apiUse ResponseLineComplete
 *
 * @apiError Unauthorized Only the connected user can access to list of lines
 * @apiError Forbidden Your are not allowed to access to this API
 *
 * @apiErrorExample {json} Unauthorized:
 *     HTTP/1.1 401 Unauthorized
 *
 * @apiErrorExample {json} Forbidden:
 *     HTTP/1.1 403 Forbidden
 */
router.route('/:lineId')
  .get((req, res) => res.json(filterLine(req.line)))
  /**
   * @api {put} /api/lines/:lineId Update the line
   * @apiName PutLine
   * @apiGroup Line
   * @apiPermission ROLE_USER
   *
   * @apiUse UpdateLine
   * @apiUse ResponseLineComplete
   */
  .put((req, res) => {
    // Force the id of the line to ensure that the user doesn't try to change it
    const data = omit(req.body, '_id')
    data._id = req.line._id

    return checkAndSaveLine(data, req, res)
  })
  /**
   * @api {delete} /api/lines/:lineId Delete the line
   * @apiName DeleteLine
   * @apiGroup Line
   * @apiPermission ROLE_USER
   *
   * @apiSuccessExample {json} Success-Response:
   *     HTTP/1.1 204 OK
   */
  .delete((req, res) => removeLine(req.line._id).then(() => res.status(204).send()))

/**
 * Filter information that the user is authorized to view in the response
 * @param {Object} line The response
 * @returns {Object} The filtered object.
 */
function filterLine (line) {
  return pick(line, '_id', 'user', 'type', 'label', 'group', 'encryption', 'updatedAt', '_rev')
}

function sanitizeInput (input) {
  const data = pick(input, '_id', 'user', 'type', 'label', 'group', 'encryption', '_rev')

  const validationError = new Error()
  validationError.status = 400

  const typeChoices = ['card', 'password', 'text']
  if (!includes(typeChoices, data.type)) {
    validationError.message = i18n.t('error:line.400.type', {choices: typeChoices.join(', ')})
    return Promise.reject(validationError)
  }

  if (!isString(data.label) || isEmpty(data.label)) {
    validationError.message = i18n.t('error:line.400.label')
    return Promise.reject(validationError)
  }

  return Promise.resolve(data)
}

/**
 * Create or modify a line given by the request.
 * @param {Object} data Object to create or modify
 * @param {Object} req Request use to read the user
 * @param {Object} res Response used to send the response.
 * @returns {Promise} Promise that the line will be saved.
 */
function checkAndSaveLine (data, req, res) {
  // Force the user stored in the line.
  data.user = req.user._id

  return sanitizeInput(data)
    .then(data => saveLine(data))
    .then(line => res.json(filterLine(line)))
}

function parseErrors (err) {
  return {
    errors: [{
      fieldName: err.field || 'global',
      message: err.message
    }]
  }
}
