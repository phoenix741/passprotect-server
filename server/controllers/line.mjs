import {pick, isString, isEmpty} from 'lodash'
import i18n from 'i18next'
import fs from 'fs'
import path from 'path'
import debug from 'debug'
import {checkPermission} from '../utils/authentification'
import {getLines, getLine, removeLine, saveLine} from '../services/line'
import {getGroups} from '../services/group'

const log = debug('App:Controllers:Line')

log('Load line type definition')
export const typeDefs = [
  fs.readFileSync(path.join(__dirname, '..', '..', 'common', 'graphql', 'line.graphql'), 'utf-8')
]

export const resolvers = {
  RootQuery: {
    async lines (obj, args, {user}) {
      checkPermission(user)
      const lines = await getLines(user)
      lines.forEach(line => checkPermission(user, line.user))
      return lines.map(filterLine)
    },

    async line (obj, {id}, {user}) {
      checkPermission(user)
      const line = await getLine(id)
      checkPermission(user, line.user)
      return filterLine(line)
    },

    async groups (obj, args, {user}) {
      checkPermission(user)
      return getGroups(user)
    }
  },
  User: {
    async lines (obj, args, {user}) {
      checkPermission(user)
      const line = await getLines(obj._id)
      checkPermission(user, line.user)

      return line.map(filterLine)
    }
  },

  RootMutation: {
    async createUpdateLine (obj, {input}, {user}) {
      log(`createUpdateLine with id ${input._id}`)
      checkPermission(user)
      try {
        if (input._id) {
          const line = await getLine(input._id)
          checkPermission(user, line.user)
          input._id = line._id
        }

        input.user = user._id

        const data = await sanitizeInput(input)
        const line = await saveLine(data)
        log(`Line will be saved with the id ${line._id}`)
        return filterLine(line)
      } catch (err) {
        return parseErrors(err)
      }
    },

    async removeLine (obj, {id}, {user}) {
      checkPermission(user)
      try {
        const line = await getLine(id)
        checkPermission(user, line.user)
        await removeLine(id)
        return {errors: []}
      } catch (err) {
        return parseErrors(err)
      }
    }
  },

  CreateUpdateLineResult: {
    __resolveType (obj) {
      if (obj.errors) {
        return 'Errors'
      }
      return 'WalletLine'
    }
  }
}

/**
 * Filter information that the user is authorized to view in the response
 * @param {Object} line The response
 * @returns {Object} The filtered object.
 */
function filterLine (line) {
  return pick(line, '_id', 'user', 'type', 'label', 'group', 'encryption', 'updatedAt', '_rev')
}

function sanitizeInput (input) {
  const data = pick(input, '_id', 'user', 'type', 'label', 'group', 'encryption', '_rev')

  const validationError = new Error()
  validationError.status = 400

  const typeChoices = ['card', 'password', 'text']
  if (!typeChoices.includes(data.type)) {
    validationError.message = i18n.t('error:line.400.type', {choices: typeChoices.join(', ')})
    throw validationError
  }

  if (!isString(data.label) || isEmpty(data.label)) {
    validationError.message = i18n.t('error:line.400.label')
    throw validationError
  }

  return data
}

function parseErrors (err) {
  return {
    errors: [{
      fieldName: err.field || 'label',
      message: err.message
    }]
  }
}
