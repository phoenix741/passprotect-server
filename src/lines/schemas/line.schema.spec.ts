import * as mongoose from 'mongoose';
import { LineSchema, migrateV1toV2 } from './line.schema';
import { LineEntity } from '../../lines/models/line.entity';
import { ObjectID } from 'bson';
import { LineTypeEnum } from '../../shared/interfaces/line-type-enum.interface';

describe('LineSchema', () => {
  let LineModel: mongoose.Model<LineEntity>;

  beforeEach(() => {
    LineModel = mongoose.model('Line', LineSchema);
  });

  describe('validation', () => {
    it('should not create a line - field required', async () => {
      expect.assertions(1);
      const line = new LineModel();
      try {
        await line.validate();
      } catch (err) {
        expect(Object.keys(err.errors)).toEqual(['user', 'group', 'type', 'label']);
      }
    });

    it('should create a line - all field fill', async () => {
      const line = new LineModel({
        label: 'Test',
        group: 'Group Test',
        type: LineTypeEnum.password,
        user: 'toto',
      });

      await line.validate();
    });
  });

  describe('migrateV1toV2', () => {
    const docV1 = {
      _id: new ObjectID('5ce1adc75fbbf35ab1af77ac'),
      type: LineTypeEnum.card,
      label: 'label',
      group: 'group',
      user: 'user',
      encryption: {
        salt: 'salt',
        informations: {
          authTag: 'authTag',
          content: 'content',
        },
      },
    };

    const docV2 = {
      _id: docV1._id,
      type: LineTypeEnum.card,
      label: 'label',
      group: 'group',
      user: 'user',
      encryption: {
        salt: 'salt',
        iv: null,
        authTag: 'authTag',
        content: 'content',
      },
    };

    const gettersSetters = {
      get(property) {
        return this[property];
      },
      set(property, value) {
        this[property] = value;
      },
    };

    it('migrateV1toV2 - migration', async () => {
      const newDoc = Object.assign({}, gettersSetters, docV1);
      migrateV1toV2((newDoc as any) as LineEntity);
      expect(newDoc).toMatchSnapshot('line migration, with migration');
    });

    it('migrateV1toV2 - no migration', async () => {
      const newDoc = Object.assign({}, gettersSetters, docV2);
      migrateV1toV2((newDoc as any) as LineEntity);
      expect(newDoc).toMatchSnapshot('line migration, without migration');
    });
  });
});
