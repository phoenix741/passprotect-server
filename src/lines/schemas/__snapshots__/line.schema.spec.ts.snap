// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineSchema migrateV1toV2 migrateV1toV2 - migration: line migration, with migration 1`] = `
Object {
  "_id": "5ce1adc75fbbf35ab1af77ac",
  "encryption": Object {
    "authTag": "authTag",
    "content": "content",
    "iv": null,
    "salt": "salt",
  },
  "get": [Function],
  "group": "group",
  "label": "label",
  "set": [Function],
  "type": "card",
  "user": "user",
}
`;

exports[`LineSchema migrateV1toV2 migrateV1toV2 - no migration: line migration, without migration 1`] = `
Object {
  "_id": "5ce1adc75fbbf35ab1af77ac",
  "encryption": Object {
    "authTag": "authTag",
    "content": "content",
    "iv": null,
    "salt": "salt",
  },
  "get": [Function],
  "group": "group",
  "label": "label",
  "set": [Function],
  "type": "card",
  "user": "user",
}
`;
